### ISSUE TITLE
Incorrect resolving of references to enum class values with shared names inside the same namespace

### ISSUE THREAD

**Issue by RockinRoel** [doxygen/doxygen#7427]
I've searched and found some maybe somewhat similar issues, but not quite the same, so I think I'll create a new issue for this.

When two scoped enums (enum class) are in the same namespace, and share an enum value name (e.g. NS::Foo::A and NS::Bar::A), then references to this common enum value A will always link to the same one, e.g. `\link NS::Bar::A \endlink` will incorrectly refer to `NS::Foo::A`.

See this example:
```cpp
/*! \namespace NS My namespace
 */
namespace NS {

/*! \enum Foo
 *  \brief Foo enum
 */
enum class Foo {
    A //!< Foo's A
};

/*! \enum Bar
 *  \brief Bar enum
 */
enum class Bar {
    A //!< Bar's A
};

/*! \class C
 */
class C {
public:
    /*! \brief Ctor
     *
     * Referencing \link Foo::A \endlink
     *
     * Referencing \link Bar::A \endlink
     */
    C();
};

}
```

When I generate the doxygen XML, I get a `classNS_1_1C.xml` containing:

```xml
        <detaileddescription>
<para>Referencing <ref refid="namespaceNS_1a1c801486f39e07df6cdff00421d7f1baa7fc56270e7a70fa81a5935b72eacbe29" kindref="member">Foo::A</ref></para>
<para>Referencing <ref refid="namespaceNS_1a1c801486f39e07df6cdff00421d7f1baa7fc56270e7a70fa81a5935b72eacbe29" kindref="member">Bar::A</ref> </para>
        </detaileddescription>
```

You can see that the `refid` is the same in both cases, both refer to `NS::Foo::A`.

I believe the culprit is the following code in `src/util.cpp` (starting from line 4277 on `master`):

```cpp
        MemberNameIterator mmli(*mn);
        MemberDef *mmd;
        for (mmli.toFirst();(mmd=mmli.current());++mmli)
        {
          const MemberDef *tmd = mmd->getEnumScope();
          //printf("try member %s tmd=%s\n",mmd->name().data(),tmd?tmd->name().data():"<none>");
          int ni=namespaceName.findRev("::");
          //printf("namespaceName=%s ni=%d\n",namespaceName.data(),ni);
          bool notInNS = tmd && ni==-1 && tmd->getNamespaceDef()==0 && (mScope.isEmpty() || mScope==tmd->name());
          bool sameNS  = tmd && tmd->getNamespaceDef() && namespaceName.left(ni)==tmd->getNamespaceDef()->name();
          //printf("notInNS=%d sameNS=%d\n",notInNS,sameNS);
          if (tmd && tmd->isStrong() && // C++11 enum class
              (notInNS || sameNS) &&
              namespaceName.length()>0  // enum is part of namespace so this should not be empty
             )
          {
            md=mmd;
            fd=mmd->getFileDef();
            gd=mmd->getGroupDef();
            if (gd && gd->isLinkable()) fd=0; else gd=0;
            //printf("Found scoped enum %s fd=%p gd=%p\n",
            //    mmd->name().data(),fd,gd);
            return TRUE;
          }
        }
```

`sameNS` becomes `true` if the enum class is in the same namespace, even if the actual name of the enum class does not match, and the `if` condition becomes true. So it will always take the first `MemberDef` in `mn` that is in the same namespace and part of a scoped enum.

It works correctly if I change the `bool sameNS = ...` line to include this check:

`... && namespaceName.mid(ni+2)==tmd->name()`

I'm not sure if this is the cleanest way to fix this, though. In any case I will make a PR with this proposed fix.

**Comment by albert-github** [comment]
Please when mentioning a `master` version please also include the hash of that moment (makes it easier to find and `master` is a moving target). In this case I think the `master` is: 9b14bf58c30a02ef19abebec280568532dc58ed4.
Also mention the method (`bool getDefs`) for easier searching.

**Comment by RockinRoel** [comment]
Ok, I'll keep that in mind if I ever report another doxygen issue.

**Comment by albert-github** [comment]
Code has been integrated in master on github (please don't close the issue as this will be done at the moment of a release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.17.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
